src/openjd/cli/_run/_local_session/_session_manager.py

```diff
@@ -112,277 +183,232 @@ def __exit__(
+                )
+            adaptive_chunk_size = max(int(adaptive_chunk_size), 1)
+            if adaptive_chunk_size != task_parameters.chunks_default_task_count:
+                LOG.info(
+                    msg=f"Open Job Description CLI: Ran {completed_task_count} tasks in {timedelta(seconds=completed_task_duration)}, average {timedelta(seconds=completed_task_duration / completed_task_count)}",
+                    extra={"session_id": self.session_id},
+                )
+                LOG.info(
+                    msg=f"Open Job Description CLI: Adjusting chunk size from {task_parameters.chunks_default_task_count} to {adaptive_chunk_size}",
+                    extra={"session_id": self.session_id},
+                )
+                task_parameters.chunks_default_task_count = adaptive_chunk_size
 
-        else:
-            # In this case, we've finished all the Tasks and exited all the environments,
-            # so we can clean up and end
-            # else:
+            # If a maximum task count was specified, count them down
+            if maximum_tasks and maximum_tasks > 0:
+                maximum_tasks -= 1
+                if maximum_tasks == 0:
+                    break
+
+    def run_step(
+        self,
+        step: Step,
+        task_parameters: Optional[Iterable[TaskParameterSet]] = None,
+        maximum_tasks: Optional[int] = None,
+    ) -> None:
+        """Run a step in the session. Optional parameters control which tasks to run."""
+        if self._openjd_session.state != SessionState.READY:
+            raise RuntimeError(
+                f"Session must be in READY state, but is in {self._openjd_session.state.name}"
+            )
+
+        LOG.info(
+            msg=f"Open Job Description CLI: Running step '{step.name}'",
+            extra={"session_id": self.session_id},
+        )
+
+        if task_parameters is None:
+            task_parameters = StepParameterSpaceIterator(space=step.parameterSpace)
+
+        # Enter all the step environments
+        self.run_environment_enters(step.stepEnvironments, EnvironmentType.STEP)
+
+        try:
+            # Run the tasks
+            if (
+                isinstance(task_parameters, StepParameterSpaceIterator)
+                and task_parameters.chunks_adaptive
+            ):
+                self._run_tasks_adaptive_chunking(step, task_parameters, maximum_tasks)
+            else:
+                # Run without adaptive chunking
+                if maximum_tasks and maximum_tasks > 0:
+                    task_parameters = islice(task_parameters, maximum_tasks)
+
+                for parameter_set in task_parameters:
+                    # This may raise a LocalSessionFailed exception
+                    self.run_task(step, parameter_set)
+        finally:
+            # Exit all the step environments
+            self.run_environment_exits(type=EnvironmentType.STEP, keep_session_running=True)
```

Comment:
Just an idea: would the environment setups and cleanups model well as a context manager? I think it'd help me understand/visualize the layers of scripts and their error handling.

Reply:
The `LocalSession` is already itself a context manager, and is handling the EXTERNAL and JOB environment types via its enter/exit methods. I'm pretty sure abstracting the code here out into another separate context manager will make things harder to follow, not easier, so I'd rather keep it as is.